### PR TITLE
Extract text from CLI to enable internationalisation

### PR DESCRIPTION
**Pull request by CarlBeek**
In order to support internationalisation, the text needs to be removed from the CLI and broken out into separate content files (`.json`).

This PR implements this via a function that determines the context in which it was called (from the call stack) and uses this information to determine the correct files to load in.

### CODE REVIEW

**Comment by samajammin** [comment]
@CarlBeek nice!

Just to confirm the file structure, to translate `eth2deposit/intl/en/cli/existing_mnemonic.json`, we'd upload the translated file e.g. for Spanish as `eth2deposit/intl/es/cli/existing_mnemonic.json` or e.g. for German `eth2deposit/intl/de/cli/existing_mnemonic.json`.

**Comment by samajammin** [comment]
@wackerow I've tested out uploading nested JSON into Crowdin & can confirm that this format is supported 👍 

File uploaded (you'll need to accept my manager invite in order to view):
https://crowdin.com/project/eth2-launchpad-test/settings#files

File translation:
https://crowdin.com/translate/eth2-launchpad-test/4/en-es?filter=basic&value=0

File: eth2deposit/utils/constants.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
 
 # File/folder constants
+INTL_CONETENT_PATH = os.path.join('eth2deposit', 'intl')
```
Comment (wackerow) [inline eth2deposit/utils/constants.py:16]: "INTL_CONTENT_PATH"?

**Comment by wackerow** [comment]
@CarlBeek This looks great! Thanks so much for tackling this.

I took a look through the codebase to find what's remaining to refactor, and will list it here to track:
- [ ] `cli/new_mnemonic.py`
- [ ] `key_handling/key_derivation/mnemonic.py`
- [ ] `settings.py`
- [ ] `deposit.py`
- [ ] `credentials.py`
- [ ] `key_handling/key_derivation/path.py`?
- [ ] `key_handling/key_derivation/tree.py`?
- [x] `cli/existing_mnemonic.py`
- [x]  `cli/generate_keys.py`
- [x] `utils/validation.py`

Happy to give a hand refactoring the remaining pages, just let me know 👍😀

File: eth2deposit/utils/constants.py (hunk 1 above)
Comment (CarlBeek) [reply eth2deposit/utils/constants.py]: Indeed, tnx. Fixed in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/b8c94a80fc8661db4affb83c13332e8e3369d92e

**Comment by CarlBeek** [comment]
@wackerow, Thanks for the list. I've tackled the remaining files in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/5fceda2e86956da588efd1e23da5bdc1b1de048b.
Regarding:
* `key_handling/key_derivation/mnemonic.py`
* `credentials.py`
* `key_handling/key_derivation/path.py`
* `key_handling/key_derivation/tree.py`
The messages and errors should never be shown to the user. They are there for internal testing & sanity checks. If a user ever sees one of those messages, it means something has gone catastrophically wrong.

`settings.py` only lists the names of the testnets & mainnet, are we planning on translating these?

File: eth2deposit/cli/new_mnemonic.py [hunk 2]
```diff
@@ -34,12 +35,13 @@ def new_mnemonic(ctx: click.Context, mnemonic_language: str, **kwargs: Any) -> N
     test_mnemonic = ''
     while mnemonic != test_mnemonic:
         click.clear()
+        # ToDo: Figure out why changing this line causes pytest to get stuck collecting.
```
Comment (CarlBeek) [inline eth2deposit/cli/new_mnemonic.py:38]: @hwwhww, any idea why using my translation function `load_text` on the following line causes pytest to hang `collecting`?

Comment (CarlBeek) [reply eth2deposit/cli/new_mnemonic.py]: Thank you so much. So simple, I just couldn't see it!

Comment (hwwhww) [reply eth2deposit/cli/new_mnemonic.py]: @CarlBeek 
Because we check the output string in the cli tests, the checks failed after you modified the string. Fixed in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/4f101752389a564f6535875b02a4dcaf8d925ee6.

I tried to use `load_text` in tests directly, but it seems to have a problem finding the path. :'(

**Comment by wackerow** [comment]
@CarlBeek Updated the list above, and looks like that testing bug was fixed? I think the only thing left is a language flag, and an initial prompt if this flag isn't provided (before selecting mnemonic language). 

Anything else I'm missing before this is ready?

- [ ] Language flag / initial prompt

**Review by hwwhww — COMMENTED** [review]
I got errors while executing the macOS binary:

```
➜ ./deposit --new-mnemonics
Traceback (most recent call last):
  File "eth2deposit/deposit.py", line 4, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "/usr/local/lib/python3.7/site-packages/PyInstaller/loader/pyimod03_importers.py", line 623, in exec_module
  File "eth2deposit/cli/existing_mnemonic.py", line 28, in <module>
  File "eth2deposit/utils/intl.py", line 40, in load_text
FileNotFoundError: [Errno 2] No such file or directory: 'eth2deposit/intl/en/cli/existing_mnemonic.json'
[51580] Failed to execute script deposit
```

I guess the `build.spec` should include the new json files to `datas`.

https://github.com/ethereum/eth2.0-deposit-cli/blob/b7f4188633ee89013eb789b6468ef2534019ae54/build_configs/macos/build.spec#L8

File: eth2deposit/intl/en/cli/generate_keys.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
+    "validate_password": {
+        "msg_password_prompt": "Type the password that secures your validator keystore(s)",
```
Comment (wackerow) [inline eth2deposit/intl/en/cli/generate_keys.json:3]: Forwarding on a comment by @samajammin on initial intl audit (#180) that I agree with... consider making this clearer that the user is about to create a new password, vs typing an existing password, eg:
`Create a password that will secure your validator keystore(s)`
"Type the password that secures" kinda implies there is already password currently securing it.

File: eth2deposit/intl/en/credentials.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{
+    "from_mnemonic": {
+        "msg_key_creation": "Creating your keys:\t\t"
```
Comment (samajammin) [inline eth2deposit/intl/en/credentials.json:3]: Is it possible for us to avoid including the tabs (`\t`) in the JSON files?

Comment (samajammin) [reply eth2deposit/intl/en/credentials.json:3]: We might be able to ignore it in Crowdin but it'd be ideal to remove these to not confuse translators.

Comment (CarlBeek) [reply eth2deposit/intl/en/credentials.json:3]: I will look into extraction this from the json and having the tabs encoded in separate print statements. 👍

**Comment by samajammin** [comment]
@CarlBeek looks like some activity here? Any update on progress? Let us know if we can help 😄

File: eth2deposit/intl/en/cli/generate_keys.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "The number of validators keys you want to generate (you can always generate more later)",
+            "prompt": "Please choose how many validators you wish to run"
+        },
+        "folder": {
+            "argument": "--folder",
+            "help": "The number of validators keys you want to generate (you can always generate more later)"
+        },
+        "chain": {
+            "argument": "--chain",
+            "help": "The version of eth2 you are targeting. Use \"mainnet\" if you are depositing ETH",
```
Comment (hwwhww) [inline eth2deposit/intl/en/cli/generate_keys.json:19]: Just realized that "the version of eth2" might be a bit off. Maybe just "The network/chain name you are targeting"?

Probably to be addressed in separate PR.

File: eth2deposit/intl/ar/cli/generate_keys.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        "err_password_mismatch": "خطأ: القيمتان اللتان تم إدخالهما غير متطابقتين. الرجاء الكتابة مرة أخرى."
+    },
+    "generate_keys_arguments_decorator": {
+        "num_validators": {
+            "argument": "--num_validators",
+            "help": "عدد مفاتيح المدققين التي تريد إنشاؤها (يمكنك دائمًا إنشاء المزيد في وقت لاحق)",
+            "prompt": "الرجاء اختيار عدد المدققين الذين ترغب في تشغيلهم"
+        },
+        "folder": {
+            "argument": "--مجلد",
```
Comment (hwwhww) [inline eth2deposit/intl/ar/cli/generate_keys.json:14]: I think the "argument" should not be translated?

```suggestion
            "argument": "--folder",
```

File: eth2deposit/intl/ar/cli/generate_keys.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "عدد مفاتيح المدققين التي تريد إنشاؤها (يمكنك دائمًا إنشاء المزيد في وقت لاحق)"
+        },
+        "chain": {
+            "argument": "--السلسلة",
```
Comment (hwwhww) [inline eth2deposit/intl/ar/cli/generate_keys.json:18]:
```suggestion
            "argument": "--chain",
```

File: eth2deposit/intl/ar/cli/existing_mnemonic.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
+{
+    "validate_mnemonic": {
+        "err_invalid_mnemonic": "هذا ليس مساعد ذاكرة صالحا، يرجى التحقق من الأخطاء المطبعية."
+    },
+    "existing_mnemonic": {
+        "arg_existing_mnemonic": {
+            "help": "إنشاء (أو استرداد) مفاتيح من مساعد ذاكرة موجود"
+        },
+        "arg_mnemonic": {
+            "argument": "--مساعد الذاكرة",
```
Comment (hwwhww) [inline eth2deposit/intl/ar/cli/existing_mnemonic.json:10]:
```suggestion
            "argument": "--mnemonic",
```

File: eth2deposit/intl/fr/cli/existing_mnemonic.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{
+    "validate_mnemonic": {
+        "err_invalid_mnemonic": "Ceci n'est pas une phrase mnémonique valide, veuillez vérifier s'il existe des fautes de frappe."
+    },
+    "existing_mnemonic": {
+        "arg_existing_mnemonic": {
+            "help": "Générer (ou récupérer) des clés à partir d'une phrase mnémonique existante"
+        },
+        "arg_mnemonic": {
+            "argument": "_mnemonic",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/existing_mnemonic.json:10]: Somehow French people like `_` over `-`? 😂
Changing back:

```suggestion
            "argument": "--mnemonic",
```

File: eth2deposit/intl/fr/cli/existing_mnemonic.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+            "help": "Phrase mnémonique que vous avez utilisée pour générer vos clés. (Il est recommandé de ne pas utiliser ce argument, et d'attendre que l'ILC vous demande votre élément mnémonique, sinon il apparaîtra dans l'historique de votre interpréteur.)",
+            "prompt": "Veuillez saisir votre phrase mnémonique en séparant les mots par des espaces (\" \")"
+        },
+        "arg_mnemonic_password": {
+            "argument": "_mnemonic_password",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/existing_mnemonic.json:15]:
```suggestion
            "argument": "--mnemonic_password",
```

File: eth2deposit/intl/fr/cli/existing_mnemonic.json [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+            "help": "Ce n'est certainement pas l'argument que vous recherchez : il est destiné aux mots de passe mnémoniques, pas aux mots de passe de keystores. Fournir un mot de passe ici sans en avoir un utilisé initialement peut entraîner la perte de clés (et donc de fonds) ! Notez également que si vous avez utilisé cet outil pour générer votre phrase mnémonique initialement, cela signifie que vous n'avez pas utilisé de mot de passe mnémonique. Cependant, si vous êtes certain d'avoir utilisé un mot de passe pour \"augmenter\" la sécurité de votre phrase mnémonique, c'est ici que vous devez le saisir."
+        },
+        "arg_validator_start_index": {
+            "argument": "_validator_start_index",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/existing_mnemonic.json:19]:
```suggestion
            "argument": "--validator_start_index",
```

File: eth2deposit/intl/fr/cli/generate_keys.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
+{
+    "validate_password": {
+        "msg_password_prompt": "Saisissez le mot de passe qui sécurise les keystores de vos validateurs.",
+        "msg_password_confirm": "Veuillez confirmer le mot de passe.",
+        "err_password_mismatch": "Erreur : Les deux valeurs saisies ne correspondent pas. Veuillez réessayer."
+    },
+    "generate_keys_arguments_decorator": {
+        "num_validators": {
+            "argument": "_num_validators",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/generate_keys.json:9]:
```suggestion
            "argument": "--num_validators",
```

File: eth2deposit/intl/fr/cli/generate_keys.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "Nombre de clés de validateurs que vous souhaitez générer (vous pourrez en générer davantage plus tard)",
+            "prompt": "Veuillez choisir le nombre de validateurs que vous souhaitez exécuter"
+        },
+        "folder": {
+            "argument": "_folder",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/generate_keys.json:14]:
```suggestion
            "argument": "--folder",
```

File: eth2deposit/intl/fr/cli/generate_keys.json [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "Nombre de clés de validateurs que vous souhaitez générer (vous pourrez en générer davantage plus tard)"
+        },
+        "chain": {
+            "argument": "_chain",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/generate_keys.json:18]:
```suggestion
            "argument": "--chain",
```

File: eth2deposit/intl/fr/cli/generate_keys.json [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "Version d'Eth2 que vous ciblez. Utilisez \"mainnet\" si vous déposez de l'ETH.",
+            "prompt": "Veuillez choisir un nom de chaîne/réseau : mainnet (réseau principal) ou testnet (réseau de test)."
+        },
+        "keystore_password": {
+            "argument": "_keystore_password",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/generate_keys.json:23]:
```suggestion
            "argument": "--keystore_password",
```

File: eth2deposit/intl/fr/cli/new_mnemonic.json [hunk 16]
```diff
@@ -0,0 +1,16 @@
+{
+    "new_mnemonic": {
+        "arg_new_mnemonic": {
+            "help": "Générer un nouvel élément mnémonique et des clés"
+        },
+        "arg_mnemonic_language": {
+            "argument": "_mnemonic_language",
```
Comment (hwwhww) [inline eth2deposit/intl/fr/cli/new_mnemonic.json:7]:
```suggestion
            "argument": "--mnemonic_language",
```

File: eth2deposit/intl/zh-CN/cli/generate_keys.json [hunk 17]
```diff
@@ -0,0 +1,35 @@
+{
+    "validate_password": {
+        "msg_password_prompt": "输入密码以保护您的验证者密钥库",
+        "msg_password_confirm": "重复以确认",
```
Comment (hwwhww) [inline eth2deposit/intl/zh-CN/cli/generate_keys.json:4]: Note for myself:

```suggestion
        "msg_password_confirm": "重复输入以确认",
```

I can create crowdin issue later.

File: eth2deposit/intl/zh-CN/cli/generate_keys.json [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "您想要生成的验证者密钥数量（您以后随时可以生成更多密钥）",
+            "prompt": "请选择您想要运行多少验证者"
+        },
+        "folder": {
+            "argument": "--folder",
+            "help": "您想要生成的验证者密钥数量（您以后随时可以生成更多密钥）"
+        },
+        "chain": {
+            "argument": "--chain",
+            "help": "您正在对准的 eth2 版本。如果您要存入以太币，请使用“主网”",
```
Comment (hwwhww) [inline eth2deposit/intl/zh-CN/cli/generate_keys.json:19]: It was wrong since "mainnet" should not be translated here:
```suggestion
            "help": "您正在对准的 eth2 版本。如果您要存入以太币，请使用 \"mainnet\"",
```

File: eth2deposit/intl/el/cli/generate_keys.json [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "Ο αριθμός των κλειδιών επαληθευτών που θέλετε να δημιουργήσετε (μπορείτε πάντα να δημιουργήσετε περισσότερα αργότερα)",
+            "prompt": "Παρακαλώ επιλέξτε πόσους επαληθευτές θέλετε να εκτελέσετε"
+        },
+        "folder": {
+            "argument": "--folder",
+            "help": "Ο αριθμός των κλειδιών επαληθευτών που θέλετε να δημιουργήσετε (μπορείτε πάντα να δημιουργήσετε περισσότερα αργότερα)"
+        },
+        "chain": {
+            "argument": "--chain",
+            "help": "Η έκδοση του eth2 στην οποία στοχεύετε. Χρησιμοποιήστε το \"κεντρικό δίκτυο\" αν κάνετε κατάθεση ETH",
```
Comment (hwwhww) [inline eth2deposit/intl/el/cli/generate_keys.json:19]: I'm not confident with my EL. But it since `mainnet` was translated here.
 
```suggestion
            "argument": "--chain",
            "help": "Η έκδοση του eth2 στην οποία στοχεύετε. Χρησιμοποιήστε το \"mainnet\" αν κάνετε κατάθεση ETH",
```

File: eth2deposit/intl/ko/cli/generate_keys.json [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+            "help": "생성하려는 검증자 키의 개수(나중에 언제든 생성 가능)",
+            "prompt": "실행하려는 검증자의 개수를 선택하세요."
+        },
+        "folder": {
+            "argument": "--folder",
+            "help": "생성하려는 검증자 키의 개수(나중에 언제든 생성 가능)"
+        },
+        "chain": {
+            "argument": "--chain",
+            "help": "타겟팅하려는 eth2의 버전입니다. ETH를 예치한다면 \"메인넷\"을 사용하세요.",
```
Comment (hwwhww) [inline eth2deposit/intl/ko/cli/generate_keys.json:19]: I guess:

```suggestion
            "argument": "--chain",
            "help": "타겟팅하려는 eth2의 버전입니다. ETH를 예치한다면 \"mainnet\"을 사용하세요.",
```

**Review by hwwhww — COMMENTED** [review]
My first scan review:
1. As we discussed offline, it's better to NOT translate the arguments
2. Some translation content issues. Can we edit here and then address to upstream Crownin later? Or we have to update Crownin and then update the JSON files?

**Review by CarlBeek — COMMENTED** [review]
Thanks for the feedback @hwwhww. I agree that CLI arguments should remain in English.

File: eth2deposit/intl/zh-CN/cli/generate_keys.json (hunk 17 above)
Comment (hwwhww) [reply eth2deposit/intl/zh-CN/cli/generate_keys.json:4]: We can merge here and then I go propose the change on Crowdin.
@wackerow what do you think?

File: eth2deposit/deposit.py [hunk 21]
```diff
@@ -1,22 +1,45 @@
...
+@click.option(
+    '--language',
+    callback=captive_prompt_callback(
+        lambda language: fuzzy_reverse_dict_lookup(language, INTL_LANG_OPTIONS),
+        choice_prompt_func(lambda: 'Please choose your language', get_first_options(INTL_LANG_OPTIONS))(),
+    ),
+    default='English',
+    prompt=choice_prompt_func(lambda: 'Please choose your language', get_first_options(INTL_LANG_OPTIONS))(),
+    required=True,
+    type=str,
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/deposit.py:39]: Added in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/0542c1d3d1b624dfed25babdbebd8d1c470919ce

File: eth2deposit/utils/intl.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+        file_path = file_path[:-3] + '.json'     # replace .py with .json
+
+    if func == '':
+        # Auto-detect function based on call stack
+        func = inspect.stack()[1].function
+
+    if lang == '':
+        lang = config.language
+
+    # Determine path to json text
```
Comment (hwwhww) [inline eth2deposit/utils/intl.py:48]: The software would break if the target string does not exist in the target language JSON file.
I think in future iterations, we might update the English string during development, and then update the language pack later. What do you think about making it:
1. Try to find the target string in the target language JSON file.
2. If it can't find, try to find the target string in the English JSON file.

**Review by hwwhww — COMMENTED** [review]
Great! Almost getting done!

Given that this PR is a milestone feature update, I think we'd only merge it once it's *ready to release*. I'd suggest that we have a pre-release so that more users can play (help debug the translations 🙈) before production release.

File: build_configs/macos/build.spec [hunk 23]
```diff
@@ -4,12 +4,15 @@ block_cipher = None
...
+             binaries=None,
+             datas=[
+                 ('../../eth2deposit/key_handling/key_derivation/word_lists/*.txt', './eth2deposit/key_handling/key_derivation/word_lists/'),
+                 ('../../eth2deposit/intl', './eth2deposit/intl',)
+             ],
              hiddenimports=[],
              hookspath=[],
              runtime_hooks=[],
-             excludes=[],
+             excludes=['FixTk', 'tcl', 'tk', '_tkinter', 'tkinter', 'Tkinter'],
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply build_configs/macos/build.spec:15]: MacOS needs this line so that the `Tk` & `Tcl` binaries don't need to be imported as per the change to [L7](https://github.com/ethereum/eth2.0-deposit-cli/pull/182/files/a7facf484c8874d6f1d3c2e37727331cfac9725e#diff-8ebc05df81d902c982839d6583795740de8efcd17ab87692907861346e6b6ab3L7) above.

I don't actually know whether PyInstaller bundles these two binaries into the Windows and Linux builds. I will investigate as removing them could shrink the final binary size a bit.

File: eth2deposit/cli/generate_keys.py [hunk 24]
```diff
@@ -78,33 +80,33 @@ def generate_keys_arguments_decorator(function: Callable[..., Any]) -> Callable[
     to obtain the necessary arguments for the generate_keys() subcommand.
     '''
     decorators = [
-        click.option(
-            '--num_validators',
-            prompt='Please choose how many validators you wish to run',
-            help='The number of validators keys you want to generate (you can always generate more later)',
+        jit_option(
+            help=lambda: load_text(['num_validators', 'help'], func='generate_keys_arguments_decorator'),
+            param_decls=lambda: load_text(['num_validators', 'argument'], func='generate_keys_arguments_decorator'),
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/cli/generate_keys.py:85]: Alright. I will strip the arguments for the JSON files and the corresponding `load_text` functions from the python files.

File: eth2deposit/utils/intl.py (hunk 22 above)
Comment (CarlBeek) [reply eth2deposit/utils/intl.py:48]: If we are missing a field accidentally, hopefully the JSON schema tests I implemented will catch that scenario (provided the schemas themselves are up to date).

Reverting to English in the case of a failure is not very intl friendly, however it does work well for ensuring that the CLI remains working. I agree with your argument around making the CLI more resilient to language changes that propagate from english to the rest of the pack.

File: eth2deposit/cli/generate_keys.py [hunk 25]
```diff
@@ -78,33 +80,33 @@ def generate_keys_arguments_decorator(function: Callable[..., Any]) -> Callable[
...
-                  'to ask you for your mnemonic as otherwise it will appear in your shell history.)'),
-            prompt='Type the password that secures your validator keystore(s)',
+            confirmation_prompt=True,
+            help=lambda: load_text(['keystore_password', 'help'], func='generate_keys_arguments_decorator'),
+            hide_input=True,
+            param_decls=lambda: load_text(['keystore_password', 'argument'], func='generate_keys_arguments_decorator'),
+            prompt=lambda: load_text(['keystore_password', 'prompt'], func='generate_keys_arguments_decorator'),
         ),
         click.option(
             '--eth1_withdrawal_address',
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/cli/generate_keys.py:112]: I don't follow. Which help string are you referring to?

File: eth2deposit/utils/intl.py (hunk 22 above)
Comment (CarlBeek) [reply eth2deposit/utils/intl.py:52]: Added English fall-over in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/3a06e92822043e56c37744fa2d34088ab9e58dbe

File: build_configs/macos/build.spec (hunk 23 above)
[the remark this answers is not in the document]
Comment (CarlBeek) [reply build_configs/macos/build.spec:15]: `Tk` and `Tcl` now both explicitly removed from Windows and Linux binaries. No noticeable difference to the size of the exported binaries, but keeping that line in there for consistency's sake.

File: eth2deposit/utils/intl.py [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+    Returns the first `option` in the values of the `options` dict.
+    '''
+    return list(map(lambda x: x[0], options.values()))
+
+
+def _closest_match(text: str, options: Iterable[str]) -> str:
+    '''
+    Finds the closest match to `text` in the `options_list`
+    '''
+    match = difflib.get_close_matches(text, options, n=1, cutoff=0.6)
```
Comment (hwwhww) [inline eth2deposit/utils/intl.py:80]: I really like how you use this method here! 👍 

Just FYI that it returns an answer even with the wrong input. e.g., input `12` gets `2: ελληνικά`.
```
Please choose your language ['1. العربية', '2. ελληνικά', '3. English', '4. Français', '5. Bahasa melayu', '6. Italiano', '7. 日本語', '8. 한국어', '9. Português do Brasil', '10. român', '11. 简体中文']:  [English]: 12
Επιλέξτε τη γλώσσα του μνημονικού σας ['1. 简体中文', '2. 繁體中文', '3. čeština', '4. English', '5. Italiano', '6. 한국어', '7. Português', '8. Español']:  [english]:
```

I think it's okay for our use case, just add a note here.

File: eth2deposit/deposit.py [hunk 27]
```diff
@@ -1,22 +1,45 @@
...
+        click.pause(load_text(['err_python_version']))
         sys.exit()
 
 
 @click.group()
-def cli() -> None:
-    pass
+@click.pass_context
+@click.option(
+    '--language',
```
Comment (hwwhww) [inline eth2deposit/deposit.py:31]: Maybe update README.md to list and explain this argument?

File: eth2deposit/cli/generate_keys.py [hunk 28]
```diff
@@ -39,23 +41,23 @@ def validate_password(cts: click.Context, param: Any, password: str) -> str:
...
             # Confirm password
-            password_confirmation = get_password(text='Repeat for confirmation')
+            password_confirmation = get_password(load_text(['msg_password_confirm']))
             if password == password_confirmation:
                 is_valid_password = True
             else:
-                click.echo('Error: the two entered values do not match. Please retype again.')
+                click.echo(load_text(['err_password_mismatch']))
 
     return password
```
Comment (hwwhww) [inline eth2deposit/cli/generate_keys.py:63]: (couldn't comment on L63 😅)

Some messages in the following `validate_eth1_withdrawal_address` also need to be added to JSON file and translated:

L67:  `raise ValueError("The given Eth1 address is not in hexadecimal encoded form.")`
and
L70: `click.echo(f'\n**[Warning] you are setting Eth1 address {normalized_address} as your withdrawal address. Please ensure that you have control over this address.**\n')`

**Review by hwwhww — APPROVED** [review]
Found some issues during QA. It looks good to me once these issues are fixed.
Awesome work! 🎉

File: eth2deposit/utils/intl.py (hunk 26 above)
Comment (CarlBeek) [reply eth2deposit/utils/intl.py:80]: My primary motivation behind this fuzzy matcher is not forcing people to have to be too precise will spelling mistakes or accents it also helps with unicode character differences (multiple ways of entering a visually similar character).

I could tune the cut-off parameter to reject this case, but the false negative rate goes up quickly then.

File: eth2deposit/cli/generate_keys.py (hunk 28 above)
Comment (CarlBeek) [reply eth2deposit/cli/generate_keys.py:71]: Resolved in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/06d79203d4d11b097fbfc5a2dafe89dc90ae97ab

File: eth2deposit/deposit.py (hunk 27 above)
Comment (CarlBeek) [reply eth2deposit/deposit.py:32]: Resolved in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/e15806bac672f26e1b214626ed2371a69325167d

File: eth2deposit/cli/generate_keys.py [hunk 29]
```diff
@@ -39,23 +41,23 @@ def validate_password(cts: click.Context, param: Any, password: str) -> str:
...
+        password = get_password(load_text(['msg_password_prompt']))
         try:
             validate_password_strength(password)
         except ValidationError as e:
-            click.echo(f'Error: {e} Please retype.')
+            click.echo(e)
         else:
             # Confirm password
-            password_confirmation = get_password(text='Repeat for confirmation')
+            password_confirmation = get_password(load_text(['msg_password_confirm']))
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/cli/generate_keys.py:64]: I've fixed this for integer range checking & `click.Choice` options in https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/b31f8142b8ca76a608d075c88291c39c62252d00 &  https://github.com/ethereum/eth2.0-deposit-cli/pull/182/commits/68bb20db4d2d373edb2b9343a66c6d5d0fcba2bd. I'll address passwords tomorrow.

**Comment by CarlBeek** [comment]
@hwwhww, I've changed quite a bit since you last looked. As far as I can see everything is working perfectlty from a translation standpoint. Please can you take one, hopefully final, look 🙏

File: eth2deposit/utils/validation.py [hunk 30]
```diff
@@ -94,6 +95,20 @@ def validate_deposit(deposit_data_dict: Dict[str, Any], credential: Credential)
...
+    '''
+    Verifies that `num` is an `int` andlow <= num < high
+    '''
+    try:
+        num_int = int(num)  # Try cast to int
+        assert num_int == float(num)  # Check num is not float
+        assert low <= num_int < high  # Check num in range
+        return num_int
+    except (ValueError, AssertionError):
+        raise ValidationError(load_text(['err_not_positive_integer']))
```
Comment (hwwhww) [inline eth2deposit/utils/validation.py:114]: It's unclear why it includes `low` but excludes `high`. What do you think about including the given upper bound and adjusting the argument:

```suggestion
def validate_int_range(num: Any, lower_bound: int, upper_bound: int) -> int::
    '''
    Verifies that `num` is an `int` and `lower_bound <= num <= upper_bound`
    '''
    try:
        num_int = int(num)  # Try cast to int
        assert num_int == float(num)  # Check num is not float
        assert lower_bound <= num_int <= upper_bound  # Check num in range
        return num_int
    except (ValueError, AssertionError):
        raise ValidationError(load_text(['err_not_positive_integer']))
```

and call `validate_int_range(num, 0, 2**32 - 1)`.

File: eth2deposit/deposit.py [hunk 31]
```diff
@@ -1,22 +1,54 @@
...
+    default='English',
+    help='The language you wish to use the CLI in.',
+    prompt=choice_prompt_func(lambda: 'Please choose your language', get_first_options(INTL_LANG_OPTIONS))(),
+    type=str,
+)
+@click.option(
+    '--non_interactive',
+    default=False,
+    is_flag=True,
+    help='Disables interactive prompts.',
```
Comment (hwwhww) [inline eth2deposit/deposit.py:46]: Unimportant, just FYI it didn't show when I execute `python eth2deposit/deposit.py  --help`. I guess it's because deposit CLI mixes `click.option` and custom `jit_option`?

**Review by hwwhww — COMMENTED** [review]
The description of the `folder` argument + the translations have to be updated; otherwise, LGTM! 👍 👍

File: eth2deposit/utils/validation.py (hunk 30 above)
Comment (CarlBeek) [reply eth2deposit/utils/validation.py:114]: Two reasons:
* It follows python's `range`/indexing convention (eg `x in range(0,8)`, `list[0,8]`)
* `validate_int_range(num, 0, 2**32)` is neater than `validate_int_range(num, 0, 2**32 - 1)`

File: eth2deposit/utils/click.py [hunk 32]
```diff
@@ -0,0 +1,116 @@
...
+    as defined by `processing_func` not returning a ValidationError
+    :param processing_func: A function to process the user's input that possibly raises a ValidationError
+    :param prompt: the text to prompt the user with, should their input raise an error when passed to processing_func()
+    :param confirmation_prompt: the optional prompt for confirming user input (the user must repeat their input)
+    :param confirmation_mismatch_msg: the message displayed to the user should their input and confirmation not match
+    :param hide_input: bool, hides the input as the user types
+    '''
+    def callback(ctx: click.Context, param: Any, user_input: str) -> Any:
+        if config.non_interactive:
+            return processing_func(user_input)
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/utils/click.py:96]: Yay indeed! It ended up being much easier than trying to coerce `click` or the script tests into doing what I wanted.

File: eth2deposit/intl/en/cli/generate_keys.json [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+        "err_invalid_ECDSA_hex_addr": "The given Eth1 address is not in hexadecimal encoded form.",
+        "msg_ECDSA_addr_withdrawal": "**[Warning] you are setting an Eth1 address as your withdrawal address. Please ensure that you have control over this address.**"
+    },
+    "generate_keys_arguments_decorator": {
+        "num_validators": {
+            "help": "The number of validators keys you want to generate (you can always generate more later)",
+            "prompt": "Please choose how many validators you wish to run"
+        },
+        "folder": {
+            "help": "The number of validators keys you want to generate (you can always generate more later)"
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply eth2deposit/intl/en/cli/generate_keys.json:12]: Aargh, I can't believe I missed this! It is correct on the master branch, it was just a copy-pasta mistake I made when extracting the text to json. That said, it it *only* in the help file, so no major loss, I will fix the English one and remove the other translations (so it will revert to explaining this in English when `--help` is requested).

File: eth2deposit/deposit.py (hunk 31 above)
Comment (CarlBeek) [reply eth2deposit/deposit.py:46]: It is because I purposefully hid this option, I intended only to use it for testing. (See [L47](https://github.com/ethereum/eth2.0-deposit-cli/pull/182/files/6f388d842ab32e239690ee56304e90952491edaa#diff-dfb2523fb295de650586cb72981df1988c32ef0749034028f84426db0a2da10aR47) for hidden argument.)

Are you of the opinion that we should be opening this up to users? (It may be useful to those integrating this into other tools)

**Comment by CarlBeek** [comment]
Thanks for your feedback, @hwwhww. I'll merge this now and we can open separate PRs to change `validate_int_range()` and `--non_interactive` if we decide this is necessary in the future.
